src/foraging_gui/GenerateMetadata.py

```diff
@@ -0,0 +1,988 @@
+        '''
+        if self.Obj['meta_data_dialog']['rig_metadata_file']=='' or self.Obj['MetadataFolder']=='':
+            return
+        
+        rig_metadata_full_path=os.path.join(self.Obj['MetadataFolder'],self.Obj['meta_data_dialog']['rig_metadata_file'])
+        with open(rig_metadata_full_path, 'w') as f:
+            json.dump(self.Obj['meta_data_dialog']['rig_metadata'], f, indent=4)
+
+    def _handle_edge_cases(self):
+        '''
+        handle edge cases (e.g. missing keys in the json file)
+        '''
+        # Missing fields camera_start_time and camera_stop_time in the Camera_dialog. 
+        # Possible reason: 1) the camera is not used in the session. 2 ) the camera is used but the start and end time are not recorded for old version of the software.
+        self._initialize_fields(dic=self.Obj['Camera_dialog'],keys=['camera_start_time','camera_stop_time'],default_value='')
+        
+        # Missing Behavior data streams in the json file.
+        # Possible reason: 1) the behavior data is not started in the session. 
+        if 'B_AnimalResponseHistory' not in self.Obj:
+            self.has_behavior_data = False
+        else:
+            self.has_behavior_data = True
+        
+        # Missing fields B_NewscalePositions, LickSpoutReferenceArea, LickSpoutReferenceX, LickSpoutReferenceY, LickSpoutReferenceZ in the json file.
+        # Possible reason: 1) the NewScale stage is not connected to the behavior GUI. 2) the session is not started.
+        if ('B_NewscalePositions' not in self.Obj) or (self.Obj['meta_data_dialog']['session_metadata']['LickSpoutReferenceArea']=='') or (self.Obj['meta_data_dialog']['session_metadata']['LickSpoutReferenceX']=='') or (self.Obj['meta_data_dialog']['session_metadata']['LickSpoutReferenceY']=='') or (self.Obj['meta_data_dialog']['session_metadata']['LickSpoutReferenceZ']==''):
+            self.has_reward_delivery = False
+        else:
+            self.has_reward_delivery = True
+
+        # Missing field WaterCalibrationResults in the json file.
+        # Possible reason: 1) the water calibration file is not included in the ForagingSettings folder. 2) the water calibration is not saved in the json file.
+        if 'WaterCalibrationResults' not in self.Obj:
+            self.Obj['WaterCalibrationResults'] = {} 
+
+        # Missing field LaserCalibrationResults in the json file.
+        # Possible reason: 1) the optogenetic calibration file is not included in the ForagingSettings folder. 2) the optogenetic calibration is not saved in the json file. 3) no optogenetics calibration for this rig.
+        if 'LaserCalibrationResults' not in self.Obj:
+            self.Obj['LaserCalibrationResults'] = {}
+
+        # Missing field open_ephys in the json file.
+        # Possible reason: 1) The ephys data is recorded but the open ephys is not controlled by the behavior GUI in the old version.
+        if 'open_ephys' not in self.Obj:
```

Comment:
What if its a behavior only session, and there is no ephys? In that case its better to just not have an `open_ephys` entry, right?

Reply:
This is a version issue. The `open_ephys` field is `[ ]` by default from the `Foraging.py` for the current version. For the old version, there is no `open_ephys` field. So, I checked the existence of the `open_ephys` in `self.Obj`.